addons/udes_stock/models/stock_quant.py

```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models, _, api
+from collections import defaultdict
+
+class StockQuant(models.Model):
+    _inherit = 'stock.quant'
+
+    def _gather(self, product_id, location_id, **kwargs):
+        """ Call default _gather function, if quant_ids context variable
+            is set the resulting quants are filtered by id.
+            Context variable quant_ids might contain quants of different products.
+            TODO: Extend default _gather to include quant_ids as kwarg
+        """
+        quants = super(StockQuant, self)._gather(product_id, location_id, **kwargs)
+        quant_ids = self.env.context.get('quant_ids')
+        if quant_ids:
+            quants = quants.filtered(lambda q: q.id in quant_ids)
+        return quants
+
+    def total_quantity(self):
+        """ Returns the total quantity of the quants in self
+        """
+        return sum(self.mapped('quantity'))
+
+    def group_quantity_by_product(self, get_key=lambda p: p.product_id.id, only_available=False):
+        """ Returns a dictionary with the total quantity per product,
+            mapped by product_id.
+            :kwargs:
+                - only_available: Boolean
+                    Whether to include those reserved or not in the grouping
+                - get_key: a callable which takes a quant and returns the key
+            :returns:
+                a dictionary with the total quantity per product,
+                    mapped by get_key or product_id.id as default
+        """
+        products = defaultdict(int)
+        for quant in self:
+            value = quant.quantity
+            if only_available:
+                value -= quant.reserved_quantity
+            products[get_key(quant)] += value
+        return products
+
+    def create_picking_from_quants(self, picking_type, **kwargs):
+        """ Create a picking from quants
+            Uses stock.picking create_picking functioanlity
+            :args:
+                - picking_type
+            :kwargs:
+                - Extra args for the create picking
+            :returns:
+                - picking
+        """
+        products = self.group_quantity_by_product(get_key=lambda p: p.product_id)
```

Comment:
`lambda p: p.product_id` is already the default `get_key`, so it is not needed.

Reply:
The default is p.product_id.id, would it be more useful to have default of  p.product_id instead?

Reply:
Also - change to q as an earlier comment suggests

Reply:
Read the messages out of order - will change to q.product_id not q.product_id.id